3_harmonize/src/clean_wqp_data.R

```diff
@@ -0,0 +1,38 @@
+#' @title Clean WQP data
+#' 
+#' @description 
+#' Function to harmonize WQP data in preparation for further analysis. Included
+#' in this function are steps to unite diverse characteristic names by assigning
+#' them to more commonly-used water quality parameter names. 
+#' 
+#' @param wqp_data data frame containing the data downloaded from the WQP, 
+#' where each row represents a unique data record.
+#' @param wqp_params list object where each element of the list is a named vector.
+#' The vectors correspond with a parameter group and contain character string(s)
+#' representing known WQP characteristic names associated with each parameter. 
+#' 
+#' @returns 
+#' Returns a formatted and harmonized data frame containing data downloaded from 
+#' the Water Quality Portal, where each row represents a unique data record.
+#' 
+clean_wqp_data <- function(wqp_data, wqp_params){
+  
+  # Before we clean any data, format a table that indicates how various 
+  # characteristic names map onto more commonly-used parameter names
+  params_df <- lapply(names(wqp_params), function(x){
+    params <- data.frame(char_name = wqp_params[[x]],
+                         parameter = x) 
+  }) %>%
+    bind_rows()
```

Comment:
I would recommend this be moved into its own target. I could see a scenario where someone might want to pass their own crosswalk of generic parameters and WQP CharacteristicNames to this function.

Reply:
Yeah, good point! I guess I'm wondering whether it'd be best to create a new target, a la `p3_param_chars_crosswalk` or to modify how we're using `p1_wqp_params`, which essentially functions as our crosswalk in this pipeline. Do you have any thoughts on that?

Reply:
Great point. I wonder if what we should do instead is pass in a CSV for `p1_wqp_params` that has two columns - `parameter_group` and `CharacteristicName`. People are generally more familiar with CSVs compared to ymls so that could be helpful. It would also mean that the "crosswalk" already exists.

Reply:
I agree that people are more familiar with CSV's, but I also personally think the yml file is much easier to read and readily interpret (by a human!), as opposed to a long-format CSV file that might contain dozens of characteristics. 

Maybe this is complicating things too much, but what if we added a new target under `p1_wqp_params` that contains the code highlighted above to build a crosswalk table, `p1_params_to_chars_crosswalk` which we could use to build `p1_char_names` with little modification. Then we could include a section in the README that says something to the effect that if you don't like the yml, just replace three of our targets (`p1_wqp_params_yml`, `p1_wqp_params`, and `p1_params_to_chars_crosswalk`) with:

```r
tar_target(
p1_wqp_params_csv,
"1_inventory/cfg/my_own_crosswalk.csv",
format = "file"
),

tar_target(
p1_params_to_chars_crosswalk
read_csv(p1_wqp_params_csv)
)
```

Reply:
Ah yes, excellent point about readability. I think the modification of 3 targets is pretty intimidating. Maybe we should keep all the code as-is, but include instructions in the README that indicate those `top-level parameter names` can be changed and the parameters within the yml moved around to customize your groupings and group names. I think the language we have in there right now discourages people from changing at all.

Reply:
OK, I recognize my yml preference might differ from other users so am still open to the csv-all-the-way option. If we keep all the code as-is, do you mean keep the "crosswalk table" formatting within `clean_wqp_data()`, or still plan on moving that code chunk to its own target?

Reply:
> I think the language we have in there right now discourages people from changing at all.

And I agree with this. At first I didn't want users to change the top-level parameter names but that shouldn't actually matter.

Reply:
> moving that code chunk to its own target

The only reason I could see needing to do this now is if we think there will ever be a need to chunk the `clean_wqp_data()` calls by using a `pattern` in `tar_target()`. I could see a national pull struggling with this step if it included all the data at once and if we did map over this target, we wouldn't want to repeat this code that didn't change between each branch.

Reply:
Thanks! I'm in favor of moving that code chunk to its own target because it 1) gives us some flexibility for changing up the cfg/crosswalk table if we later choose to do that, and 2) potentially makes the pipeline more readily scalable for the reasons you point out above.